### DISCUSSION TITLE
Filamat VS Matc

### DISCUSSION THREAD

**Discussion by ThomasGorisse** [google/filament#4118] [Q&A] [answered]
Hi,

I'd like to know if there is a serious gain using a matc generated .filamat comparing to doing it at runtime on the device with Filamat for small materials like this:
```
vec3 desaturate(vec3 color, float amount) {
    vec3 gray = vec3(dot(vec3(0.2126, 0.7152, 0.0722), color));
    return vec3(mix(color, gray, amount));
}

void material(inout MaterialInputs material) {
    prepareMaterial(material);
    vec4 color = texture(materialParams_videoTexture, getUV0()).rgba;
    vec3 keyColor = materialParams.chromaKeyColor.rgb;
    float threshold = 0.675;
    float slope = 0.2;
    float distance = abs(length(abs(keyColor - color.rgb)));
    float edge0 = threshold * (1.0 - slope);
    float alpha = smoothstep(edge0, threshold, distance);
    color.rgb = desaturate(color.rgb, 1.0 - (alpha * alpha * alpha));

    material.baseColor.a = alpha;
    material.baseColor.rgb = inverseTonemapSRGB(color.rgb);
    material.baseColor.rgb *= material.baseColor.a;
}
```
Since Sceneform already has a good number of matc generated files, I'd like not to increase the Filament version upgrade time with more .mat to .filamat by just having to update the Filamat-Lite dependency version.

Thanks

**Comment by romainguy** [comment, marked as answer]
`matc` is just a command line wrapper around `filamat`. However, `filamat-lite` has two major limitations: it can only target GLSL (this means it cannot be used with Vulkan or Metal backends), and does not perform any optimizations. Not having optimizations is probably ok but they are greatly helpful for compatibility as they simplify the generated shader.

That said, why is running `matc` an issue? You only need to recompile material files when we change the material version, which happens rarely nowadays (I think we did once or twice at most in the last 12 months or so).

**Reply by ThomasGorisse** [reply]
OK, you convinced me. We will go for matc even more because shaders could be a performance and debugging nightmare if not well written and I don't want to see issues coming from wrong usages.

One more question: Does an exception is thrown [here](https://github.com/google/filament/blob/122908587b103c84b980905c8ef47802a3dd337a/android/filament-android/src/main/java/com/google/android/filament/Material.java#L318) in case of .filamat incompatibility ?

**Reply by romainguy** [reply]
No, I believe we just crash with an assert when the version doesn't match. It makes it very obvious that you haven't upgraded your material files :)

**Comment by ThomasGorisse** [comment]
That makes the update checks quick.

Thanks
